### PR TITLE
Add test cases Validator stats in Superfluid module

### PR DESCRIPTION
**Pull request by hieuvubk**
Closes: #2177

## What is the purpose of the change

> Add test cases for `Unbonded`, `Unbonding` validator stat

### CODE REVIEW

File: x/superfluid/keeper/distribution_test.go [hunk 1]
```diff
@@ -55,6 +55,14 @@ func (suite *KeeperTestSuite) TestMoveSuperfluidDelegationRewardToGauges() {
 			[]int64{0},
 			[]gaugeChecker{{0, 0, 0, true}, {1, 1, 0, false}},
 		},
+		{
```
Comment (mattverse) [inline x/superfluid/keeper/distribution_test.go:58]: can we add more detailed comments for what we're expecting from this test case?
e.g) unbonding validators are not expected to be rewarded

File: x/superfluid/keeper/hooks_test.go [hunk 2]
```diff
@@ -283,6 +303,13 @@ func (suite *KeeperTestSuite) TestBeforeSlashingUnbondingDelegationHook() {
 				suite.Require().NoError(err)
 				// slash by slash factor
 				power := sdk.TokensToConsensusPower(validator.Tokens, sdk.DefaultPowerReduction)
+
+				// should not be slashing unbonded validator
```
Comment (mattverse) [inline x/superfluid/keeper/hooks_test.go:307]: asking to learn : does it not matter for the unbonding case?

File: x/superfluid/keeper/intermediary_account_test.go [hunk 3]
```diff
@@ -26,6 +26,18 @@ func (suite *KeeperTestSuite) TestIntermediaryAccountCreation() {
 			2,
 			[]superfluidDelegation{{0, 0, 0, 1000000}, {1, 1, 0, 1000000}},
 		},
+		{
```
Comment (mattverse) [inline x/superfluid/keeper/intermediary_account_test.go:29]: nit: can we add details on what we're testing in the name or comments for the expected behavior when the validator is unbonded or unbonding?

File: x/superfluid/keeper/slash_test.go [hunk 4]
```diff
@@ -25,6 +25,22 @@ func (suite *KeeperTestSuite) TestBeforeValidatorSlashed() {
...
+		{
+			"with an unbonding validator and single superfluid delegation",
+			[]stakingtypes.BondStatus{stakingtypes.Unbonding},
+			1,
+			[]superfluidDelegation{{0, 0, 0, 1000000}},
+			[]int64{0},
+			[]int64{0},
+		},
+		{
+			"with single validator and single superfluid delegation",
```
Comment (mattverse) [inline x/superfluid/keeper/slash_test.go:37]: Should be changed to unbonded!

**Review by mattverse — COMMENTED** [review]
LGTM overall! Nice work figuring out all the cases for unbonding and unbonded cases must have been gnarly. Would you mind adding comments for what we're expecting for when it is unbonding or unbonded? I think also adding documentation on the method itself on how stuff work when validators are unbonded & unbonding would be also super helpful!

File: x/superfluid/keeper/hooks_test.go (hunk 2 above)
Comment (hieuvubk) [reply x/superfluid/keeper/hooks_test.go:307]: Yes i tested it does not matter for the unbonding case?

File: x/superfluid/keeper/distribution_test.go (hunk 1 above)
Comment (hieuvubk) [reply x/superfluid/keeper/distribution_test.go:61]: It was actually not allocating the reward to the Unbonded validator. Therefore gauges are not distributed.
I added the case of allocating to both Unbonded and Unbonding vals and gauges are distributed normally.
Also in these tests, except for the case that slash does not work with Unbonded validator. I think there is no difference between the 3 stats together

File: x/superfluid/keeper/distribution_test.go [hunk 5]
```diff
@@ -47,9 +47,20 @@ func (suite *KeeperTestSuite) TestMoveSuperfluidDelegationRewardToGauges() {
...
+		// There is no difference between Bonded, Unbonding, Unbonded
 		{
-			"add unbonded validator case",
-			[]stakingtypes.BondStatus{stakingtypes.Bonded, stakingtypes.Unbonded},
+			"add unbonded, unbonding validator case",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded, stakingtypes.Unbonded, stakingtypes.Unbonding},
+			3,
+			[]superfluidDelegation{{0, 0, 0, 1000000}, {1, 1, 0, 1000000}, {2, 2, 0, 1000000}},
+			[]int64{0,1,2},
+			[]gaugeChecker{{0, 0, 0, true}, {1, 1, 0, true}, {2, 2, 0, true}},
```
Comment (czarcas7ic) [inline x/superfluid/keeper/distribution_test.go:58]: Sorry I don't quite understand this. Why are we expecting a reward in the unbonding case here, but in the next test we do not expect a reward for the same unbonding case? I'm not saying you are wrong, it's just this is unclear to me.

File: x/superfluid/keeper/intermediary_account_test.go [hunk 6]
```diff
@@ -26,6 +26,19 @@ func (suite *KeeperTestSuite) TestIntermediaryAccountCreation() {
 			2,
 			[]superfluidDelegation{{0, 0, 0, 1000000}, {1, 1, 0, 1000000}},
 		},
+		// Can create intermediary account with unbonded, unbonding validators
+		{
+			"test intermediary account with unbonded validator",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded, stakingtypes.Unbonded},
+			2,
+			[]superfluidDelegation{{0, 0, 0, 1000000}, {1, 1, 0, 1000000}},
```
Comment (czarcas7ic) [inline x/superfluid/keeper/intermediary_account_test.go:34]: Is there a reason we allow intermediary account creation with unbonded validators? Seems odd to me, but I might be missing something.

File: x/superfluid/keeper/slash_test.go [hunk 7]
```diff
@@ -25,6 +25,22 @@ func (suite *KeeperTestSuite) TestBeforeValidatorSlashed() {
...
+		},
+		{
+			"with an unbonded validator and single superfluid delegation",
+			[]stakingtypes.BondStatus{stakingtypes.Unbonded},
+			1,
+			[]superfluidDelegation{{0, 0, 0, 1000000}},
+			[]int64{0},
+			[]int64{0},
+		},
 		// {
```
Comment (czarcas7ic) [inline x/superfluid/keeper/slash_test.go:44]: Should we remove this dead code?

File: x/superfluid/keeper/intermediary_account_test.go (hunk 6 above)
Comment (czarcas7ic) [reply x/superfluid/keeper/intermediary_account_test.go:34]: This is just an aside, no change required on this PR in regards to this!

File: x/superfluid/keeper/distribution_test.go (hunk 5 above)
Comment (hieuvubk) [reply x/superfluid/keeper/distribution_test.go:58]: In the next test, I want to show that If there is no reward to the validator, the gauge will not be distributed.

Comment (czarcas7ic) [reply x/superfluid/keeper/distribution_test.go:58]: Ohh, the `rewardVals` field confused me, I now understand

**Review by czarcas7ic — APPROVED** [review]

**Review by mattverse — APPROVED** [review]
🌮

**Pull request merged**